vdc/scripts/users/environment.yml

```diff
@@ -0,0 +1,10 @@
+name: users
+dependencies:
+  - python=3.7
+  - pip
+  - pip:
+      - kubernetes
+      - google-api-python-client
+      - google-cloud-storage
+      - shortuuid
```

Comment:
I thought about truncation; from my reading, this may not be safe, lose more uniqueness than you'd expect by generating a random string of equal length. As a trivial example: https://devblogs.microsoft.com/oldnewthing/20080627-00/?p=21823

The shortuuid code uses the full uuid v4 int representation, modulo desired length N to generate an index into the dictionary, then repeats on the quotient until remainder is 0. This, I believe, guarantees maximum entropy (for a string length N), provided that your dictionary equals you're characterset. I slightly truncate the dictionary, since I need to make sure that the first and last entry is alphanumeric. 
* https://github.com/skorokithakis/shortuuid/blob/master/shortuuid/main.py#L16

@patrick-schultz may know more about this.

Reply:
I don't know much about uuid, but if I understand the shortuuid code correctly, it's converting the uuid to a string in a lossless way, which doesn't make it any shorter in actual bits. It's really just converting the uuid to a base b representation, where b is the length of the alphabet.

What are you using shortuuid for?

Reply:
Service account, bucket id creation.

Reply:
So is the purpose just to convert the uuid into a string representation, not to make a shorter uuid?

Reply:
String and shorten. Maximum length is 30 characters.

Reply:
I did the math. As long as the alphabet is at least 20 characters, a 128-bit uuid will always convert to a string no longer than 30 characters. Doesn't look like we need to worry about truncation.

Reply:
I see, thanks.

I think the math is 2^128 = x^r ; for r = 30 characters , x = 20 unique characters (base 20), and so an alphabet [a-z0-9], or even [a-z0-3] is sufficient.

Hex is 16 characters, so needs 32 characters, because 128 / log2(16) = 32. This is why uuid's hex digest is 32 characters long (ignoring the dashes) for this reason.

shortuuid's conversion function looks like this (pseudo):

```
number = uuid.v4().int
alphabet = ["a", "b" ... ]
base = len(alphabet)

result = []
while number:
    number, remainder = divmod(number, base)
    result.append(alphabet[remainder])
```

I think this is a [typical way to change bases](http://faculty.bucks.edu/cornelld/numbersystem.htm#10%20to%20base%202,%208,%20or%2016%20use%20the%20remainder). So provided that the alphabet used is as you say at least 20 unique values, shortuuid's output will be as random as uuid v4, and be no longer than 30 characters.

So based on this, it seems better to continue using shortuuid vs truncating uuidv4 to 30 characters.

@patrick-schultz, @jigold, thank you for the feedback and discussion! Does this proposal seem ok?

Reply:
👍

Reply:
Great! @jigold [it isn't included in mega-docker requirements.txt](https://github.com/hail-is/hail/blob/master/docker/requirements.txt). Should I add it?

Reply:
Yes, I think so. It's not breaking the tests right now because you're not running any tests for this project. I think to add tests, you need a `hail-ci-build.sh` file and the project added to `projects.yaml`. Any package you depend on needs to be in the mega docker file requirements file in `docker/requirements.txt`. I think that's the only place it needs to go. Some projects still have conda in them like pipeline and ci. That will take time to convert them from not using conda.

Reply:
Why not install it on top of the base image in our own Dockerfile. We had spoken about installing dependencies in the mega-docker only if they were used by 2 or more packages.